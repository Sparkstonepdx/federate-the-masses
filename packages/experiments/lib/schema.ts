import set from 'lodash-es/set';
import { Schema } from './core-record-types';

interface SchemaEnrichment {
  references?: string[];
  referencedBy?: Record<string, boolean>;
}

export class SchemaEngine {
  constructor(private schema: Record<string, Schema & SchemaEnrichment>) {
    this.computeRelationships();
  }

  computeRelationship(key: string, visitedKeys = new Set()) {
    if (visitedKeys.has(key)) return []; //this.schema[key].references ?? [];
    let references: string[] = [];
    visitedKeys.add(key);
    let schema = this.schema[key];
    if (!schema) throw new Error(`Schema doesn't exist: ${key}`);

    for (const [field, fieldSchema] of Object.entries(schema.fields)) {
      if (fieldSchema.type !== 'relation') continue;
      let nestedReferences = this.computeRelationship(fieldSchema.collection, visitedKeys).map(
        ref => [fieldSchema.collection, ref].join('.'),
      );

      set(this.schema, [key, 'referencedBy', fieldSchema.collection], true);

      if (nestedReferences.length) {
        references.push(...nestedReferences);
      } else {
        references.push(fieldSchema.collection);
      }
    }
    this.schema[key].references = references;
    return references;
  }

  computeRelationships(keys?: string[], visitedKeys = new Set()) {
    keys ??= Object.keys(this.schema);

    for (const key of keys) {
      this.computeRelationship(key, visitedKeys);
    }
  }

  get(collectionName: string) {
    return this.schema[collectionName];
  }
}
