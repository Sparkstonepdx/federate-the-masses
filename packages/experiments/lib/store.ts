export interface RecordData {
  id: string;
  is_deleted?: boolean;
  modified_at: string;
  created_at: string;
}

export interface Store {
  get(collectionName: string, id: string): Promise<RecordData | undefined>;
  set(collectionName: string, id: string, value: RecordData): Promise<void>;
  list(CollectionName: string): Promise<RecordData[]>;
  delete(collectionName: string, id: string): Promise<void>;
}

export class MemoryStore implements Store {
  public db = new Map<string, Map<string, RecordData>>();

  constructor(initialData?: object) {
    if (initialData) {
      for (const [collectionName, records] of Object.entries(
        initialData ?? {},
      )) {
        const recordMap = new Map();
        for (const [id, record] of Object.entries(records ?? {})) {
          recordMap.set(id, record);
        }

        this.db.set(collectionName, recordMap);
      }
    }
  }

  get(collection: string, id: string) {
    return Promise.resolve(this.db.get(collection)?.get(id));
  }

  set(collection: string, id: string, value: RecordData) {
    if (!this.db.has(collection)) this.db.set(collection, new Map());
    this.db.get(collection)!.set(id, value);
    return Promise.resolve();
  }

  list(collection: string) {
    return Promise.resolve(Array.from(this.db.get(collection)?.values() ?? []));
  }

  delete(collection: string, id: string) {
    this.db.get(collection)?.delete(id);
    return Promise.resolve();
  }
}
