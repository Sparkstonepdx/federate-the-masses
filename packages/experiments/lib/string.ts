import { omit } from 'lodash-es';
import { Record } from './records';

export function prettyPrint<RecordType extends object = {}>(record?: Record<RecordType> | null) {
  if (!record) return `[] ${record}`;
  let data = omit(record.data(), 'created_at', 'modified_at', 'id');

  let prettyData = Object.entries(data).map(([key, value]) => `${key}: ${JSON.stringify(value)}`);

  return `[${record.collection}:${record.id}] ${prettyData.join(', ')}`;
}

export function prettyPrintArray(arr: Record[]) {
  return arr.map(prettyPrint).join('\n');
}
