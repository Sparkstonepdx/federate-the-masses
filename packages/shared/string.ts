import { omit, pick } from 'lodash-es';
import { CollectionRecord } from '../server/lib/records';

export function prettyPrint<RecordType extends object = {}>(
  record?: CollectionRecord<RecordType> | null,
  pickFields?: string[]
) {
  if (!record) return `[] ${record}`;
  let data: any = omit(record.data(), 'created_at', 'modified_at', 'id');
  if (pickFields) {
    data = pick(record.data(), pickFields);
  }

  let prettyData = Object.entries(data).map(([key, value]) => `${key}: ${JSON.stringify(value)}`);

  return `[${record.collection}:${record.id}] ${prettyData.join(', ')}`;
}

export function prettyPrintArray(arr: CollectionRecord[]) {
  return arr.map(record => prettyPrint(record)).join('\n');
}

export async function prettyList(
  promise: Promise<{ records: CollectionRecord[] }>,
  pickFields?: string[]
) {
  const result = await promise;
  return {
    ...result,
    records: result.records.map(record => prettyPrint(record, pickFields)),
  };
}
