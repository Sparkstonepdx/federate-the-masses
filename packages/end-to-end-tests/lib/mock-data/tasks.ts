import { Schema } from '../../../shared/core-record-types';
import systemSchema from '../../../shared/system-schema';

/*
    list-1 (Project Alpha) 5 items (task-3 twice)
    ├── task-1: Design homepage
    │   └── task-2: Develop homepage
    │       └── task-3: Review homepage   (belongs to list-2)
    └── list-2 (Sublist of Alpha)
        └── task-3: Review homepage       (same as above)

    list-3 (Project Beta)
    └── task-4: Setup database
*/

export const schema: Record<string, Schema> = {
  ...systemSchema,
  tasks: {
    collectionName: 'tasks',
    fields: {
      list: { type: 'relation', collection: 'lists', required: true },
      parent: { type: 'relation', collection: 'tasks' },
      child: { type: 'relation', collection: 'tasks', via: 'parent' },
      title: { type: 'string' },
      content: { type: 'string', required: true },
    },
  },
  lists: {
    collectionName: 'lists',
    fields: {
      list: { type: 'relation', collection: 'lists' },
      tasks: { type: 'relation', collection: 'tasks', via: 'list' },
      child_list: { type: 'relation', collection: 'lists', via: 'list' },
      title: { type: 'string' },
    },
  },
};

export const data = {
  lists: {
    'list-1': {
      id: 'list-1',
      title: 'Project Alpha',
    },
    'list-2': {
      id: 'list-2',
      title: 'Sublist of Alpha',
      list: 'list-1',
    },
    'list-3': {
      id: 'list-3',
      title: 'Project Beta',
    },
  },
  tasks: {
    'task-1': {
      id: 'task-1',
      title: 'Design homepage',
      content: 'Create wireframes and mockups',
      list: 'list-1',
    },
    'task-2': {
      id: 'task-2',
      title: 'Develop homepage',
      content: 'Convert design into code',
      list: 'list-1',
      parent: 'task-1',
    },
    'task-3': {
      id: 'task-3',
      title: 'Review homepage',
      content: 'Internal QA and feedback',
      list: 'list-2',
      parent: 'task-2',
    },
    'task-4': {
      id: 'task-4',
      title: 'Setup database',
      content: 'Install PostgreSQL and configure',
      list: 'list-3',
    },
  },
};
