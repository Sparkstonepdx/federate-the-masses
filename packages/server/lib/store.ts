import { RecordPage, Schema, SchemaField } from '../../shared/core-record-types';

export interface RecordData {
  id: string;
  is_deleted?: boolean;
  modified_at: string;
  created_at: string;
  host: string;
}

export interface FindOptions {
  filter?: string;
  page?: number;
  perPage?: number;
  sort?: string;
  fields?: string;
  expand?: string[];
}

export interface Store {
  get<Record>(collectionName: string, id: string): Promise<(Record & RecordData) | undefined>;
  set(collectionName: string, id: string, value: RecordData): Promise<void>;
  list<Record>(CollectionName: string): Promise<RecordPage<Record & RecordData>>;
  find<Record>(
    collectionName: string,
    options: FindOptions
  ): Promise<RecordPage<Record & RecordData>>;
  delete(collectionName: string, id: string): Promise<void>;
  close(): Promise<void>;

  createCollection(collectionName, fields: Record<string, SchemaField>): Promise<void>;
  addFields(collectionName: string, schema: Record<string, SchemaField>): Promise<void>;
  removeFields(collectionName: string, fields: string[]): Promise<void>;
  addIndex(collectionName: string, indexName: string, fields: string[]): Promise<void>;
  removeIndex(collectionName: string, indexName: string): Promise<void>;

  // add these for completeness:

  changeFieldType(collectionName: string, field: string, newSchema: SchemaField): Promise<void>;

  // addConstraint?(coll: string, constraintName: string, sql: string): Promise<void>; // e.g. FOREIGN KEY, CHECK
  // removeConstraint?(coll: string, constraintName: string): Promise<void>;

  // renameIndex?(coll: string, oldName: string, newName: string): Promise<void>;

  dropTable?(coll: string): Promise<void>;
}
