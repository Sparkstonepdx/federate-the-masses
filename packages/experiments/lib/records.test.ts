import { expect, test, vi } from 'vitest';
import Server from './server';
import { MemoryStore } from './store';
import { SchemaEngine } from './schema';
import { data, schema } from './mock-data/tasks';
import { prettyPrint, prettyPrintArray } from './string';

test('expand single forward relation', async () => {
  const server = new Server({
    store: new MemoryStore(data),
    schema: new SchemaEngine(schema),
    identity: { url: 'http://test-server.com', public_key: 'asdf' },
  });

  const list = await server.records.get('lists', 'list-2');

  if (!list) throw new Error('failed to get list');
  await server.records.expand(list, ['list']);

  expect(prettyPrint(list.expand.list)).toMatchInlineSnapshot(
    `"[lists:list-1] title: "Project Alpha""`
  );
});

test('expand via relation', async () => {
  const server = new Server({
    store: new MemoryStore(data),
    schema: new SchemaEngine(schema),
    identity: { url: 'http://test-server.com', public_key: 'asdf' },
  });

  const list = await server.records.get('lists', 'list-2');

  if (!list) throw new Error('failed to get list');
  await server.records.expand(list, ['tasks']);

  expect(prettyPrintArray(list.expand.tasks)).toMatchInlineSnapshot(
    `"[tasks:task-3] title: "Review homepage", content: "Internal QA and feedback", list: "list-2", parent: "task-2""`
  );
});

test('expand all tasks', async () => {
  const server = new Server({
    store: new MemoryStore(data),
    schema: new SchemaEngine(schema),
    identity: { url: 'http://test-server.com', public_key: 'asdf' },
  });

  const list = await server.records.get('lists', 'list-1');

  if (!list) throw new Error('failed to get list');
  await server.records.expand(list, ['tasks.child.child', 'child_list']);

  expect(list).toMatchSnapshot();
});

test('expand all tasks with overlap', async () => {
  const server = new Server({
    store: new MemoryStore(data),
    schema: new SchemaEngine(schema),
    identity: { url: 'http://test-server.com', public_key: 'asdf' },
  });

  const list = await server.records.get('tasks', 'task-1');

  if (!list) throw new Error('failed to get list');
  await server.records.expand(list, ['child.child', 'child.parent']);

  expect(list).toMatchSnapshot();
});
