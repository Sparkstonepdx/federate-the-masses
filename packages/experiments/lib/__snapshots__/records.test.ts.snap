// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expand all tasks 1`] = `
{
  "collection": "lists",
  "data": {
    "id": "list-1",
    "title": "Project Alpha",
  },
  "expand": {
    "child_list": [
      {
        "collection": "lists",
        "data": {
          "id": "list-2",
          "list": "list-1",
          "title": "Sublist of Alpha",
        },
        "expand": undefined,
        "id": "list-2",
      },
    ],
    "tasks": [
      {
        "collection": "tasks",
        "data": {
          "content": "Create wireframes and mockups",
          "id": "task-1",
          "list": "list-1",
          "title": "Design homepage",
        },
        "expand": {
          "child": [
            {
              "collection": "tasks",
              "data": {
                "content": "Convert design into code",
                "id": "task-2",
                "list": "list-1",
                "parent": "task-1",
                "title": "Develop homepage",
              },
              "expand": {
                "child": [
                  {
                    "collection": "tasks",
                    "data": {
                      "content": "Internal QA and feedback",
                      "id": "task-3",
                      "list": "list-2",
                      "parent": "task-2",
                      "title": "Review homepage",
                    },
                    "expand": undefined,
                    "id": "task-3",
                  },
                ],
              },
              "id": "task-2",
            },
          ],
        },
        "id": "task-1",
      },
      {
        "collection": "tasks",
        "data": {
          "content": "Convert design into code",
          "id": "task-2",
          "list": "list-1",
          "parent": "task-1",
          "title": "Develop homepage",
        },
        "expand": {
          "child": [
            {
              "collection": "tasks",
              "data": {
                "content": "Internal QA and feedback",
                "id": "task-3",
                "list": "list-2",
                "parent": "task-2",
                "title": "Review homepage",
              },
              "expand": {
                "child": [],
              },
              "id": "task-3",
            },
          ],
        },
        "id": "task-2",
      },
    ],
  },
  "id": "list-1",
}
`;

exports[`expand all tasks 2`] = `
{
  "collection": "lists",
  "data": {
    "id": "list-1",
    "title": "Project Alpha",
  },
  "expand": {
    "tasks": [
      {
        "collection": "tasks",
        "data": {
          "content": "Create wireframes and mockups",
          "id": "task-1",
          "list": "list-1",
          "title": "Design homepage",
        },
        "expand": {
          "child": [
            {
              "collection": "tasks",
              "data": {
                "content": "Convert design into code",
                "id": "task-2",
                "list": "list-1",
                "parent": "task-1",
                "title": "Develop homepage",
              },
              "expand": {
                "child": [
                  {
                    "collection": "tasks",
                    "data": {
                      "content": "Internal QA and feedback",
                      "id": "task-3",
                      "list": "list-2",
                      "parent": "task-2",
                      "title": "Review homepage",
                    },
                    "expand": {
                      "parent": {
                        "collection": "tasks",
                        "data": {
                          "content": "Convert design into code",
                          "id": "task-2",
                          "list": "list-1",
                          "parent": "task-1",
                          "title": "Develop homepage",
                        },
                        "expand": undefined,
                        "id": "task-2",
                      },
                    },
                    "id": "task-3",
                  },
                ],
              },
              "id": "task-2",
            },
          ],
        },
        "id": "task-1",
      },
      {
        "collection": "tasks",
        "data": {
          "content": "Convert design into code",
          "id": "task-2",
          "list": "list-1",
          "parent": "task-1",
          "title": "Develop homepage",
        },
        "expand": {
          "child": [
            {
              "collection": "tasks",
              "data": {
                "content": "Internal QA and feedback",
                "id": "task-3",
                "list": "list-2",
                "parent": "task-2",
                "title": "Review homepage",
              },
              "expand": {
                "child": [],
              },
              "id": "task-3",
            },
          ],
        },
        "id": "task-2",
      },
    ],
  },
  "id": "list-1",
}
`;

exports[`expand all tasks with overlap 1`] = `
{
  "collection": "tasks",
  "data": {
    "content": "Create wireframes and mockups",
    "id": "task-1",
    "list": "list-1",
    "title": "Design homepage",
  },
  "expand": {
    "child": [
      {
        "collection": "tasks",
        "data": {
          "content": "Convert design into code",
          "id": "task-2",
          "list": "list-1",
          "parent": "task-1",
          "title": "Develop homepage",
        },
        "expand": {
          "child": [
            {
              "collection": "tasks",
              "data": {
                "content": "Internal QA and feedback",
                "id": "task-3",
                "list": "list-2",
                "parent": "task-2",
                "title": "Review homepage",
              },
              "expand": undefined,
              "id": "task-3",
            },
          ],
          "parent": {
            "collection": "tasks",
            "data": {
              "content": "Create wireframes and mockups",
              "id": "task-1",
              "list": "list-1",
              "title": "Design homepage",
            },
            "expand": undefined,
            "id": "task-1",
          },
        },
        "id": "task-2",
      },
    ],
  },
  "id": "task-1",
}
`;
